lib/ansible/module_utils/network/aci/msc.py

```diff
@@ -0,0 +1,299 @@
+        else:
+            self.module.fail_json(msg="Parameter 'password' is required for authentication")
+
+    def login(self):
+        ''' Log in to MSC '''
+
+        # Perform login request
+        self.url = urljoin(self.baseuri, 'auth/login')
+        payload = {'username': self.params['username'], 'password': self.params['password']}
+        resp, auth = fetch_url(self.module,
+                               self.url,
+                               data=json.dumps(payload),
+                               method='POST',
+                               headers=self.headers,
+                               timeout=self.params['timeout'],
+                               use_proxy=self.params['use_proxy'])
+
+        # Handle MSC response
+        if auth['status'] != 201:
+            self.response = auth['msg']
+            self.status = auth['status']
+            self.fail_json(msg='Authentication failed: {msg}'.format(**auth))
+
+        payload = json.loads(resp.read())
+
+        self.headers['Authorization'] = 'Bearer {token}'.format(**payload)
+
+    def request(self, path, method=None, data=None):
+        ''' Generic HTTP method for MSC requests. '''
+        self.path = path
+
+        if method is not None:
+            self.method = method
+
+        self.url = urljoin(self.baseuri, path)
+        resp, info = fetch_url(self.module,
+                               self.url,
+                               headers=self.headers,
+                               data=json.dumps(data),
+                               method=self.method,
+                               timeout=self.params['timeout'],
+                               use_proxy=self.params['use_proxy'],
+                               )
+        self.response = info['msg']
+        self.status = info['status']
+
+        # 200: OK, 201: Created, 202: Accepted, 204: No Content
+        if self.status in (200, 201, 202, 204):
+            output = resp.read()
+            if self.method in ('DELETE', 'PATCH', 'POST', 'PUT') and self.status in (200, 201, 204):
+                self.result['changed'] = True
+            if output:
+                return json.loads(output)
+
+        # 404: Not Found
+        elif self.method == 'DELETE' and self.status == 404:
+            return {}
+
+        # 400: Bad Request, 401: Unauthorized, 403: Forbidden,
+        # 405: Method Not Allowed, 406: Not Acceptable
+        # 500: Internal Server Error, 501: Not Implemented
+        elif self.status >= 400:
+            try:
+                payload = json.loads(resp.read())
+            except:
+                payload = json.loads(info['body'])
+            if 'code' in payload:
+                self.fail_json(msg='MSC Error {code}: {message} [{info}]'.format(**payload), payload=data)
+            else:
+                self.fail_json(msg='MSC Error:'.format(**payload), info=info, output=output)
+
+        return {}
+
+    def query_objs(self, path, **kwargs):
```

Comment:
I see you have query_objs and get_obj. What is the difference between query and get?

Reply:
The function `get_obj()` uses `query_obj()`, so they are quite similar.

The difference is purely whether you are querying a single object, or multiple objects. If you expect only a single object, `get_obj()` is for you, if you expect multiple objects `query_objs()` is for you.

Both accept filter arguments, but some things should always return a single obj (like if name or id is unique). The only reason we have this function call is purely to ensure that when we expect a single object, and we do get more than one, we escalate to the user.